lib/std/sort.zig

```diff
@@ -427,375 +427,388 @@ test "sort fuzz testing" {
-    );
-    try testing.expectEqual(
-        @as(?usize, 1),
-        binarySearch(R, @as(i32, -20), &[_]R{ R.r(-100, -50), R.r(-40, -20), R.r(-10, 20), R.r(30, 40) }, {}, R.order),
-    );
-}
-
-/// Returns the index of the first element in `items` greater than or equal to `key`,
-/// or `items.len` if all elements are less than `key`.
+
+    try std.testing.expectEqual(null, binarySearch(u32, &[_]u32{}, @as(u32, 1), S.orderU32));
+    try std.testing.expectEqual(0, binarySearch(u32, &[_]u32{1}, @as(u32, 1), S.orderU32));
+    try std.testing.expectEqual(null, binarySearch(u32, &[_]u32{0}, @as(u32, 1), S.orderU32));
+    try std.testing.expectEqual(null, binarySearch(u32, &[_]u32{1}, @as(u32, 0), S.orderU32));
+    try std.testing.expectEqual(4, binarySearch(u32, &[_]u32{ 1, 2, 3, 4, 5 }, @as(u32, 5), S.orderU32));
+    try std.testing.expectEqual(0, binarySearch(u32, &[_]u32{ 2, 4, 8, 16, 32, 64 }, @as(u32, 2), S.orderU32));
+    try std.testing.expectEqual(1, binarySearch(i32, &[_]i32{ -7, -4, 0, 9, 10 }, @as(i32, -4), S.orderI32));
+    try std.testing.expectEqual(3, binarySearch(i32, &[_]i32{ -100, -25, 2, 98, 99, 100 }, @as(i32, 98), S.orderI32));
+    try std.testing.expectEqual(null, binarySearch(R, &[_]R{ R.r(-100, -50), R.r(-40, -20), R.r(-10, 20), R.r(30, 40) }, @as(i32, -45), R.order));
+    try std.testing.expectEqual(2, binarySearch(R, &[_]R{ R.r(-100, -50), R.r(-40, -20), R.r(-10, 20), R.r(30, 40) }, @as(i32, 10), R.order));
+    try std.testing.expectEqual(1, binarySearch(R, &[_]R{ R.r(-100, -50), R.r(-40, -20), R.r(-10, 20), R.r(30, 40) }, @as(i32, -20), R.order));
+    try std.testing.expectEqual(2, binarySearch([]const u8, &[_][]const u8{ "", "abc", "1234", "vwxyz" }, @as(usize, 4), S.orderLength));
+}
+
+/// Returns the index of the first element in `items` returning `.eq` or `.gt`
+/// when given to `compareFn`.
+/// - Returns `items.len` if all elements return `.lt`.
+///
+/// `items` must be sorted in ascending order with respect to `compareFn`:
+/// ```
+/// [0]                                                   [len]
+/// ┌───┬───┬─/ /─┬───┬───┬───┬─/ /─┬───┬───┬───┬─/ /─┬───┐
+/// │.lt│.lt│ \ \ │.lt│.eq│.eq│ \ \ │.eq│.gt│.gt│ \ \ │.gt│
+/// └───┴───┴─/ /─┴───┴───┴───┴─/ /─┴───┴───┴───┴─/ /─┴───┘
+/// ├─────────────────┼─────────────────┼─────────────────┤
+///  ↳ zero or more    ↳ zero or more    ↳ zero or more
+///                   ├───┤
+///                    ↳ returned index
+/// ```
 ///
-/// `items` must be sorted in ascending order with respect to `compareFn`.
+/// `O(log n)` time complexity.
 ///
-/// O(log n) complexity.
+/// See also: `binarySearch`, `upperBound`, `partitionPoint`, `equalRange`.
 pub fn lowerBound(
     comptime T: type,
-    key: anytype,
     items: []const T,
     context: anytype,
-    comptime lessThan: fn (context: @TypeOf(context), lhs: @TypeOf(key), rhs: T) bool,
+    comptime compareFn: fn (@TypeOf(context), T) std.math.Order,
```

Comment:
why did you switch this from lessThan to compare, and then wrap it to make it back into lessThan?